lib/Sema/CSSolver.cpp

```diff
@@ -1402,6 +1402,27 @@ tryOptimizeGenericDisjunction(ConstraintSystem &cs, Constraint *disjunction,
       return nullptr;
   }
 
+  // Don't attempt this optimization if call has number literals.
+  // This is intended to narrowly fix situations like:
+  //
+  // func test<T: FloatingPoint>(_: T) { ... }
+  // func test<T: Numeric>(_: T) { ... }
+  //
+  // test(42)
+  //
+  // The call should use `<T: Numeric>` overload even though the
+  // `<T: FloatingPoint>` is a more specialized version because
+  // selecting `<T: Numeric>` doesn't introduce non-default literal
+  // types.
```

Comment:
Shockingly (to me), I just tried this in present-day Swift, and in fact it **does** select the `<T: FloatingPoint>` overload:

```swift
func test<T: FloatingPoint>(_: T) { print("Hi") }
func test<T: Numeric>(_: T) { print("Bye") }

test(42) // "Hi"
```

Reply:
Yeah, this is due to a hack that compares two (!) generic overloads without accounting for arguments, add a third and you'd get `<T: Numeric>` selected i.e.:

```swift
func test<T: FloatingPoint>(_: T) { print("Hi") }
func test<T: Numeric>(_: T) { print("Bye") }
func test(_: String) {}

test(42)
```

Because the hack is no longer applied.